Kafka Connect Sink - to send messages from Kafka to IoT devices

varunpuranik opened the pull request:
Added support for a Kafka Connect Azure IoT Hub Sink that sends messages from Kafka topics to IoT devices. The messages need to be in a particular format as specified in the documentation. This does not yet include sending feedback from the sent messages back to some Kafka topic.

varunpuranik committed:
Move Kafka Connect IoT Hub Source files to Source folder

To prep the repo to add code for IoT Hub Sink Connector, moving all the source connector code to source folder.

varunpuranik committed:
Add support for IoT Hub Sink Connector

Adding code for a Sink Connector that reads messages from Kafka topics
and sends them to IoT devices via IoT Hub.
Also adding tests for the Sink connector.

varunpuranik committed:
Add README for Sink Connector

Added README for the Kafka Sink Connector. Also split the main README
into 2 - one for the sink and one for the source.

varunpuranik committed:
Update README_Sink with additional info

Added information about each of the fields that the Kafka messages need
to have.

varunpuranik committed:
Add support for messages in JSON format

- Added support for messages in JSON serialized format
- Added connect-iothub-sink.properties template file
- Updated README and Tests.

varunpuranik committed:
Update README for Sink and handle wait timeout

- Updated README for Sink cover cases when a SchemaRegistry is being
used and when it is not being used.
- Also handled the case when it takes too long for the Flush to wait
for the send tasks to complete.

msftclas commented:
Hi __@varunpuranik__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Varun Puranik). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

dluc commented on README.md:
```diff
@@ -1,190 +1,25 @@
+is used to pump data from [Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/) to
+[Apache Kafka](https://kafka.apache.org/), whereas the sink connector reads messages from Kafka and sends them to IoT
+ devices via [Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/). When used in tandem, the 2
+ connectors allow communicating with IoT devices by simply posting and reading messages to Kafka topics. This should
```
possible typo: "posting and reading messages to/from ..."

dluc commented on README.md:
```diff
@@ -1,190 +1,25 @@
+[Apache Kafka](https://kafka.apache.org/), whereas the sink connector reads messages from Kafka and sends them to IoT
+ devices via [Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/). When used in tandem, the 2
+ connectors allow communicating with IoT devices by simply posting and reading messages to Kafka topics. This should
+ makie it easier for open source systems and other systems that already interface with Kafka to communicate with
```
typo "makie"

dluc commented on README_Sink.md:
```diff
@@ -0,0 +1,182 @@
+
+The Sink connector allows you to send messages to Azure IoT devices by simply posting messages to Kafka topics. The
+connector sends the messages to Azure IoT Hub, which in turn forwards them to the right devices. The
+messages need to be in a specific format (details below), that allow the connector to extract the information
```
typo "allows"

dluc commented on README_Sink.md:
````diff
@@ -0,0 +1,182 @@
+```
+
+> Note: At the moment, the Sink connector does not support getting feedback from the devices on the status of the
+> messages (whether accepted or rejected). If you want to get feedback, you will need to do it manually. Please refer
````
accepted/rejected/expired

dluc commented on README_Sink.md:
```diff
@@ -0,0 +1,182 @@
+
+The sink connector expects the messages in the Kafka topic to have the schema below, so that it can extract the
+information required to send the message to the IoT device. If a record is not in the expected format, the connector
+will throw an exception.
```
perhaps we should clarify if the connector stops or continues to work, ignoring the poison message

dluc commented on README_Source.md:
```diff
@@ -0,0 +1,179 @@
+tasks.max=1
+Kafka.Topic=IotTopic
+IotHub.EventHubCompatibleName=iothub-toketi
+IotHub.EventHubCompatibleEndpoint=sb://iothub-ns-toketi-001.servicebus.windows.net/
```
here I would use a more generic name, e.g. sb://my-iothub-001.servicebus.windows.net/

dluc commented on README_Source.md:
```diff
@@ -0,0 +1,179 @@
+IotHub.EventHubCompatibleName=iothub-toketi
+IotHub.EventHubCompatibleEndpoint=sb://iothub-ns-toketi-001.servicebus.windows.net/
+IotHub.AccessKeyName=service
+IotHub.AccessKeyValue=6XdRSFB9H61f+N3uOdBJiKwzeqbZUj1K//T2jFyewN4=
```
please make sure that this key and the others in the PR, are not valid

varunpuranik replied:
Yes, they are all dummy.

dluc commented on build.sbt:
```diff
@@ -19,6 +19,7 @@ libraryDependencies ++= {
   val scalaTestVersion = "3.0.0"
   val configVersion = "1.3.1"
   val json4sVersion = "3.5.0"
+  val iotHubServiceClientVersion = "1.0.11"
```
the latest is 1.0.12, with a new namespace+name

"com.microsoft.azure.sdk.iot" % "iot-service-client" % "1.0.12"

dluc commented on src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessage.scala:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.Date
+
+case class C2DMessage(messageId: String, message: String, deviceId: String, expiryTime: Option[Date] = None)
```
we could support the ACK modes, see com.microsoft.azure.sdk.iot.service.sdk.DeliveryAcknowledgement: None, NegativeOnly, PositiveOnly, Full

varunpuranik replied:
These are supported as config values.

dluc commented on src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/C2DMessageConverter.scala:
```diff
@@ -0,0 +1,110 @@
+/*
```
nitpick: please keep public methods before the private ones

dluc commented on src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubMessageSender.scala:
```diff
@@ -0,0 +1,24 @@
+    this.serviceClient.sendAsync(deviceId, message)
+  }
+
+  def close(): Unit = {
```
could this return a Future too, like the previous method?

varunpuranik replied:
Changes this to use the sync version, as the async versions has some bugs.

dluc commented on src/main/scala/com/microsoft/azure/iot/kafka/connect/sink/IotHubSinkConnector.scala:
```diff
@@ -0,0 +1,53 @@
+
+  override def taskConfigs(maxTasks: Int): util.List[util.Map[String, String]] = {
+    val configList = (1 to maxTasks).map(c => this.props.asJava).toList.asJava
+    configList
```
nitpick: you could avoid the local val and use a single line of code

dluc reviewed (approved):
Some minor changes it would be nice to have before releasing

varunpuranik committed:
Addressing PR feedback and changing using sync version of ServiceClient.sync

- Several minor updates to address PR feedback
- Changed to use the sync version of ServiceClient.sendMessage (as the
async one has some bugs)
- Made some updates to the README.

varunpuranik committed:
Update README to use release 0.6

Created a new release which contains both the source and sink
connectors and updated the READMEs to point to the right release.
